mcrypt-2.6.4/doc/mcrypt.1
.PU
.TH MCRYPT 1 "29 May 2002" "local"
.SH NAME
mcrypt, mdecrypt \- encrypt or decrypt files
.SH SYNOPSIS
.ll +8
.B mcrypt
.RB [ " \-dLFubhvrzp " ]
.RB [ \-a\ algorithm ]
.RB [ "\-c\ config_file" ]
.RB [ \-m\ mode ]
.RB [ \-s\ keysize ]
.RB [ \-o\ keymode ]
.RB [ "\-k\ key1 key2 ..." ]
.RB [ \-f\ keyfile ]
[
.I "filename \&..."
]
.ll -8
.br

.B mdecrypt
.RB [ " \-LFusbhvzp " ]
.RB [ \-a\ algorithm ]
.RB [ "\-c\ config_file" ]
.RB [ \-m\ mode ]
.RB [ \-s\ keysize ]
.RB [ \-o\ keymode ]
.RB [ "\-k\ key1 key2 ..." ]
.RB [ \-f\ keyfile ]
[
.I "filename \&..."
]
.br
.SH DESCRIPTION
.I Mcrypt
is a simple crypting program, a replacement for the old unix 
.IR crypt "(1)".
When encrypting or decrypting a file, a new file is created with the
extension
.B "\&.nc"
and mode 0600. The new file keeps the modification date of the original.
The original file may be deleted by specifying the -u
parameter.
If no files are specified, the standard input is encrypted to the standard 
output.
.PP

.I Mcrypt
uses all the symmetric algorithms included in libmcrypt.

.SH HINTS
By default,
.I mcrypt
, when one of these algorithms is specified, prompts something like:
 Enter passphrase: ...
 You should then enter a passphrase long enough (512 characters is the
maximum length). Now in order to encrypt the file, The passphrase is 
transformed using the specified (or the default) key generation algorithm, 
and a random salt. The produced value is then used as the key, which is fed to the 
algorithm.
.PP
.B Algorithm Vulnerability:
Most algorithms today are designed to resist in specific attacks.
None of them is proved not to be vulnerable to some kind of attack not
as yet known.
.PP
.B Compression:
By compressing your data before encryption you gain
both in efficiency (faster encryption) and safety of your data (language redundancy
is removed). A drawback is that most compression programs will add specific
headers in the compressed file, thus making known plaintext attacks easier.
Compression after encryption is useless and may result to compressed files with longer size than the original.
.PP
.B Error Recovery:
There is some error recovery in mcrypt. 
If bytes are removed or lost from the file or stream in ECB, CBC and OFB modes,
are impossible to recover, although CFB mode will recover. If some
bytes are altered then a full block of plaintext is affected in ECB mode,
two blocks in CBC and CFB modes, but only the corresponding byte in OFB mode.
Mcrypt uses a 32 bit CRC to check for errors in the encrypted files.
.PP
.B Extra security:
For the very paranoid, if mcrypt is executed with superuser priviledges it
ensures that no important data (keys etc.) are written to disk, as swap etc.
Keep in mind that mcrypt was not designed to be a setuid program, so you
shouldn't make it one.
.PP 
Do not rely on the fact that an algorithm has a large key size, try to use
long passphrases and try to make them unpredictable.
.PP

All the block algorithms above support these modes of encryption:

.B ECB: 
The Electronic CodeBook mode. It is the simplest mode to use with a 
block cipher. Encrypts each block independently.

.B CBC:
The Cipher Block Chaining mode. It is better than ECB since the plaintext
is XOR'ed with the previous ciphertext. A random block is placed as the
first block so the same block or messages always encrypt to something
different. (This is the default mode)

.B CFB:
The Cipher-Feedback Mode (in 8bit). This is a self-synchronizing
stream cipher implemented from a block cipher. 

.B OFB:
The Output-Feedback Mode (in 8bit). This is a synchronous
stream cipher implemented from a block cipher. It is intended for use
in noisy lines, because corrupted ciphertext blocks do not corrupt the
plaintext blocks that follow. Insecure (because used in 8bit mode) so I recommend against
using it. Added just for completeness.

.B nOFB:
The Output-Feedback Mode (in nbit). n Is the size of the block of the
algorithm. This is a synchronous stream cipher implemented from a block
cipher. It is intended for use in noisy lines, because corrupted ciphertext
blocks do not corrupt the plaintext blocks that follow. 

.PP
Encrypted files can be restored to their original form using
.I mcrypt -d
or
.I mdecrypt
.
.PP
.I mdecrypt
takes a list of files on its command line and creates a new file for each
file whose name ends with .nc by removing the ".nc" or by adding ".dc"
to the end of the file name if .nc is not in the encrypted file's name.

.SH OPTIONS
.TP
.B \-F --force
Force output on standard output or input from stdin if that is a terminal.
By default mcrypt will not output encrypted data to terminal, nor read
encrypted data from it.
.TP
.B \-z --gzip
Use gzip (if it exists in your system) to compress files before encryption.
If specified at decryption time it will decompress these files. 
.TP
.B \-p --bzip2
Use bzip2 (if it exists in your system) to compress files before encryption.
If specified at decryption time it will decompress these files. 
.TP
.B \ --openpgp-z INT
This option will enable compression in OpenPGP (RFC2440) encrypted files.
.TP
.B \-d --decrypt
Decrypt.
.TP
.B \ --help
Display a help screen and quit.
.TP
.B \-v --version
Version. Display the version number and quit.
.TP
.B \-L --license
Display the
.I mcrypt's
license and quit.
.TP
.B \-o --keymode MODE
MODE may be one of the keymodes listed by the --list-keymodes parameter.
It actually is the convertion to the key before it is fed to the algorithm.
It is recommended to leave it as is, if you do not know what it is.
However if you still want to use this option, you might want to
use the 'hex' mode which allows you to specify the key in hex
(and no convertion will by applied).
.TP
.B \-h --hash HASH_ALGORITHM
HASH_ALGORITHM may be one of the algorithms listed by the --hashlist parameter.
This is the digest that will be appended to the file to be encrypted,
in order to detect file corruption. The default is the CRC32 checksum.
.TP
.B \-s --keysize SIZE
SIZE is the algorithm's key size in bytes (not the size of the passphrase). It
defaults to the maximum key supported by the algorithm. The maximum key sizes of
the algorithms may be obtained by the --list parameter. It is safe not to
touch this.
.TP
.B \-g --openpgp
This option will make mcrypt to use the OpenPGP (RFC2440) file format
for encrypted files. This will make files encrypted by mcrypt accessible
from any OpenPGP compliant application.
.TP
.B \-b --bare
No important information like the algorithm, mode, the bit mode and
the crc32 of the original file are written in the encrypted file. 
The security lies on the algorithm not on obscurity so this is NOT the 
default. This flag must also be specified when decrypting a bare encrypted 
file. 
When the bare flag is specified decryption and encryption are faster. This
may be usefull when using mcrypt to encrypt a link or something like that.
.TP
.B --flush
Flushes the output (ciphertext or plaintext) immediately. Usefull if mcrypt
is used with pipes.
.TP
.B --time
Prints some timing information (encryption speed etc.)
.TP
.B --nodelete
When this option is specified mcrypt does not delete the output file, even
if decryption failed. This is usefull if you want to decrypt a corrupted
file.
.TP
.B \-q --quiet
Suppress some not critical warnings.
.TP
.B \-u --unlink
Unlink (delete) the input file if the whole process of encryption/decryption
succeeds. This is not the default in order to use an external program
to remove sensitive data.
.TP
.B \ --list
Lists all the algorithms currently supported.
.TP
.B \ --list-keymodes 
Lists all the key modes currently supported.
.TP
.B \ --hashlist
Lists all the hash algorithms currently supported.
.TP
.B \-r --random
Use /dev/(s)random instead of /dev/urandom. This may need some key input
or mouse move to proceed. 
If your system does not support /dev/random or /dev/urandom, a 
random gatherer will be used.
.TP
.B \-k --key KEY1 KEY2 ...
Enter the keyword(s) via the command line. The KEY(s) is/are then used as
keyword instead of prompting for them. Keep in mind that someone
may see the command you are executing and so your keyword(s).
.TP
.B \-c --config FILE
Use the specified configuration file. The default is .mcryptrc
in your home directory. The format of the configuration
file is the same as the parameters. An example file is:
 algorithm safer+
 mode cbc
 key a_very_secret_one

.TP
.B \-f --keyfile FILE
Enter the keyword(s) via a file. One keyword is read per line.
The first keyword read is used for the first file, the second for
the second file etc. If the keywords are less than the files then
the last keyword is used for the remaining. A limitation is that you cannot
use the NULL (\\0) and the Newline (\\n) character in the key.
A solution to this problem is to specify the keyword in hex mode.
.TP
.B \-m --mode MODE
Mode of encryption and decryption. These modes are currently supported:
ECB, CFB, OFB, nOFB, CBC and STREAM. CBC is the default. Unless the bare flag 
is specified there is no need to specify these modes for decryption.
For stream algorithms (like WAKE) mode should be STREAM.
.TP
.B \-a --algorithm ALGORITHM
The algorithm used to encrypt and decrypt. Unless the bare flag is specified
there is no need to specify these for decryption.

The algorithms currently supported are shown with the --list parameter.
.fi
.PP
.SH "EXAMPLES"
For mcrypt to be compatible with the solaris 
.IR des "(1)",

mcrypt-2.6.4/NEWS
21Feb2003: Version 2.6.4
	- Updated Spanish-Argentina translation by Ariel
	Fermani <the_end@bbs.frc.utn.edu.ar>.  
	- Some fixes for compiling under win32.
        - Some fixes to allow compiling gaaout.c with external
          CFLAGS and CPPFLAGS.

04Oct2002: Version 2.6.3
	- Added support for 64 bit file offsets. Based on patch by
          Keven Belanger <kbelanger@logicon.ca>.

23Jun2002: Version 2.6.2
	- Corrections in localization
	- Added Spanish-Argentina translation. Translated by Ariel
	Fermani <the_end@bbs.frc.utn.edu.ar>.  

15Jun2002: Version 2.6.1
	- Better error checking
	- Added rndunix random gatherer from gnupg. It is
	  a gatherer for random bytes, written by Peter Gutmann.
	- Added some kind of random byte generator for Win32
	  systems.
	- Corrected bug in configuration file parsing.
	- Corrected bug in bare mode which put an IV in the 
	  encrypted file even if the mode did not support IV.

29May2002: Version 2.6.0
	- Added OpenPGP support (added by Timo Schulz <ts@winpt.org>)
	- Removed all of file locking code.
	- Several improvements and corrections on the old 
          codebase (still a mess).

11Mar2002: Version 2.5.13
	- Corrected stream modes in block algorithms

29Jan2002: Version 2.5.12
	- Added some missing files

26Jan2002: Version 2.5.11 
	- SHA1 is the default digest used

01Dec2001:
	- Added --time option

24Sep2001: Version 2.5.10
	- Corrected Makefile
	- Corrected bug in key checking

15Sep2001: Version 2.5.9
	- Corrected bug which made mcrypt block while writing the header
	  (patch by Jerome Bertorelle <jerome.bertorelle@noos.fr>)
	- Several cleanups

08Sep2001: Version 2.5.8
	- Added option for staticaly linking mcrypt
	- '--gzip' and '--bzip2' options now work even when handling streams
	  (based on patch by Stefan Hetzl <shetzl@teleweb.at>)
	- Removed libmcrypt-nm and libmcrypt 2.2 support
	- Updated magic file (for file(1) program)
	- Updated included gettext

04Jun2001: Version 2.5.7 released
	- Fixes for the new libmcrypt 

27Apr2001: Version 2.5.6 released
	- Updated locales 

02Oct2000: Version 2.5.5 released.
	- Support for libmcrypt linked against libdl
	- Some more verbose error messages

19May2000: Version 2.5.4 released.
	- Support for the libmcrypt-nm.
	- bugfixes in argument handling

02Apr2000: Version 2.5.3 released.
	- mcrypt does not depend now in the new libmhash. It can work
	  with the older version as well.

21Mar2000: Version 2.5.2 released.
	- Adapted to the libmcrypt 2.4 API
	- The unix crypt mode parameters were changed.
	  In order to use it, you should use the parameters:
          "-a crypt --bare" AND "--keymode scrypt" (instead of "--keymode asis").
	- The '-' file name is now used to refer to stdin.

08Mar2000: Version 2.5.1 released.
	- Mcrypt now checks the keysizes before asking for a passphrase.
	- Corrected bugs in command line arguments proccessing.

06Mar2000: Version 2.5.0 released.
	- The way crc32 is stored into the encrypted file has changed.
        - Now mcrypt supports hash algorithms.
        - Updated manpage.
	- The option --nodelete was added. If specified it will not delete
          the output file, even if crc32 check fails.
	- Mcrypt can now compile using the 2.2.5 version of libmcrypt.
          That way it looses some of it's functionality.
        - Mcrypt distribution no longer includes libmcrypt.
	- New (extendable) File format. Moved to version 2.5.0 since not all planned
	  improvements were added (to justify the 3.0.0 version).

19Jan2000: Corrected a minor bug in src/extra.c, where bzero used
               sizeof() for malloc'ed data.

10Dec1999: Corrected a bug in src/mcrypt.c, encrypt_general() function.
               It did not use the random number generator properly for
               stream ciphers.

03Dec1999: Corrected a serious bug in src/extra.c, read_iv function.
               It didn't work when not using streams.
           Corrected a bug in mcrypt.c, function encrypt_general(), which 
               made the program to crash when a smaller than the maximum 
               keysize was specified at the command line.
           Documentation was updated.
           Bugfix in bare decryption for block algorithms. When a file was 
               smaller than a fixed number it was not decrypted as it should. 

02Dec1999: Mcrypt can now compile in systems were libintl.h was not found
               without disabling nls, patch by Sascha Schumann.
28Nov1999:
           --keygen_directory parameter was replaced by --keymode_directory.

26Nov1999:
           --algorithms_directory, --modes_directory, --keygen_directory options added.
               the user can now specify where the libmcrypt modules are.


mcrypt-2.6.4/src/mcrypt.c
/*  mcrypt 2.5 - encrypts text or binary files using symmetric algorithms.
 *  For a brief description of the algorithms read the man page.
 *
 *    Copyright (C) 1998,1999,2000,2001,2002 Nikos Mavroyanopoulos
 *
 *    This program is free software; you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation; either version 2 of the License, or
 *    (at your option) any later version.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with this program; if not, write to the Free Software
 *    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */


/* $Id: mcrypt.c,v 1.36 2002/12/18 11:37:27 nmav Exp $ */

#ifndef DEFINES_H
#define DEFINES_H
#include <defines.h>
#endif
#include <mcrypt_int.h>
#include <xmalloc.h>
#include <extra.h>
#include <keys.h>
#include <random.h>
#include <gaa.h>
#include <errors.h>
#include <environ.h>
#include <popen.h>
#include <openpgp.h>

#ifdef HAVE_FTIME
# include <sys/timeb.h>
#else
# include <time.h>
#endif

static char rcsid[] =
    "$Id: mcrypt.c,v 1.36 2002/12/18 11:37:27 nmav Exp $";

char tmperr[128];
unsigned int stream_flag = FALSE;
char *keymode = NULL;
char *mode = NULL;
char *algorithm = NULL;
char *hash = NULL;
int keysize;
int nodelete = FALSE;		/* Delete by default */
int noiv = FALSE;
int hash_algorithm = MHASH_CRC32;
int noecho = TRUE;		/* Echo asterisks by default */
int double_check = FALSE;	/* Do not double check for passwords when decrypting */
int quiet = TRUE;		/* silent by default */
int tmpi;
int unlink_flag = FALSE;
int bare_flag = FALSE, real_random_flag = FALSE;
int cleanDelete = FALSE;
#ifdef ZIP
int gzipflag = FALSE;
int bzipflag = FALSE;
#endif
int flush = FALSE;
int timer = FALSE;
int openpgp = FALSE;
int openpgp_z = 0; /* default is no compression */
char *outfile = 0, *tmpc;
char *algorithms_directory = NULL;
char *modes_directory = NULL;
char* program_name = NULL;

void usage(void)
{
	gaa_help();
   fprintf(stdout, _("\n"
		     "Report bugs to mcrypt-dev@lists.hellug.gr.\n\n"));
	exit(0);

}

static char* remove_suffix( char* filename);
void mcrypt_license()
{
   fprintf(stdout,
	   _("\nCopyright (C) 1998-2002 Nikos Mavroyanopoulos\n"
	     "This program is free software; you can redistribute it and/or modify \n"
	     "it under the terms of the GNU General Public License as published by \n"
	     "the Free Software Foundation; either version 2 of the License, or \n"
	     "(at your option) any later version. \n" "\n"
	     "This program is distributed in the hope that it will be useful, \n"
	     "but WITHOUT ANY WARRANTY; without even the implied warranty of \n"
	     "MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the \n"
	     "GNU General Public License for more details. \n" "\n"
	     "You should have received a copy of the GNU General Public License \n"
	     "along with this program; if not, write to the Free Software \n"
	     "Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.\n\n"));
}

int check_hash_algo(char *chain)
{
   int i;
   char *y;
   char tmp[255];

   for (i = 0; i < 255; i++) {
      y = mhash_get_hash_name(i);
      if (y != NULL) {
	 strcpy(tmp, y);
	 free(y);
	 mcrypt_tolow(tmp, strlen(tmp));
	 if (strcmp(tmp, chain) == 0) {
	    return i;
	 }
      }
   }
   return -1;
}


#ifndef LIBMCRYPT_VERSION
# ifdef LIBMCRYPT24
#  define LIBMCRYPT_VERSION "2.4"
# endif
#endif

void mcrypt_version()
{

   fprintf(stderr, _("Mcrypt v.%s (%s-%s-%s)\n"), MCRYPT_VERSION, T_CPU,
	   T_VENDOR, T_OS);
   fprintf(stderr, _("Linked against libmcrypt v.%s\n"),
	   mcrypt_check_version(NULL));
   fprintf(stderr,
	   _
	   ("Copyright (C) 1998-2002 Nikos Mavroyanopoulos (nmav@gnutls.org)\n"));

}

/* Checks if the given key size is supported by the algorithm
 */
static void check_keysize( int keysize) {
      int *siz, i, y, x;

      if ( keysize >
	  mcrypt_module_get_algo_key_size(algorithm,
					  algorithms_directory)) {
	 err_quit(_
		  ("The specified key size is too large for this algorithm.\n"));
      }

      siz =
	  mcrypt_module_get_algo_supported_key_sizes(algorithm,
						     algorithms_directory,
						     &i);
      x = 0;
      if (siz == NULL || i == 0)
	 x = 1;			/* we have already checked if it is
				   * less than the maximum supported
				   * by the algorithm
				 */
      else {
	 for (y = 0; y < i; y++) {
	    if (siz[y] == keysize) {
	       x = 1;
	       break;
	    }
	 }
	 free(siz);
      }

      if (x == 0) {
	 err_quit(_
		  ("The specified key size not valid for this algorithm.\n"));
      }
}


/*                              The main    */
int main(int argc, char **argv)
{

   short int ein = 0, din = 0, kin = 0, force = 0, return_val = 0;
   char *einfile = 0, *dinfile = 0, *keyword = 0;
   char **file, **keyfile = NULL, *cfile;
   int x, y, i, file_count = 0, keys = 0, used_algo = FALSE;
   gaainfo info;

#ifdef HAVE_SIGNAL_H
   Signal(SIGINT, shandler);
   Signal(SIGQUIT, shandler);
   Signal(SIGSEGV, shandler);
   Signal(SIGPIPE, shandler);
   Signal(SIGTERM, shandler);
   Signal(SIGHUP, shandler);
   Signal(SIGUSR1, SIG_IGN);
   Signal(SIGUSR2, SIG_IGN);
   Signal(SIGALRM, SIG_IGN);
#endif

#ifdef ENABLE_NLS
   setlocale(LC_ALL, "");
   bindtextdomain(PACKAGE, LOCALEDIR);
   textdomain(PACKAGE);
#endif

   program_name = argv[0];
   
   if ((gaa(argc, argv, &info)) != -1) {
      err_quit(_
	       ("Error in the arguments. Use the --help parameter for more info.\n"));
   }
   if (info.config_file != NULL) {
      cfile = _mcrypt_strdup( info.config_file);
      if (check_file(cfile) != 1) {
	 err_crit(_("Warning: Cannot access config file.\n"));
      }
   } else {
/*
#ifdef HAVE_GETPWUID
      cfile = get_cfile(getuid(), ".mcryptrc");
#else*/
      cfile = _mcrypt_strdup( ".mcryptrc");
//#endif
   }

   /* If config file exists 
    */
   if (check_file(cfile) == 1) {
      if ((gaa_file(cfile, &info)) != -1) {
	 /* gets options from file 'config' */
	 err_quit(_
		  ("Error in the configuration file. Use the --help parameter for more info.\n"));
      }
   }
   _mcrypt_free(cfile);


   if (check_env() == TRUE) {

      if (get_env_key() != NULL) {
	 info.keylen = 1;
	 _mcrypt_free(info.keys);
	 info.keys = get_env_key();
      }
      if (get_env_algo() != NULL) {
	 _mcrypt_free(info.algorithm);
	 info.algorithm = get_env_algo();
      }
      if (get_env_mode() != NULL) {
	 _mcrypt_free(info.mode);
	 info.mode = get_env_mode();
      }
      if (get_env_bit_mode() != 0) {
	 info.kmode = get_env_bit_mode();
      }
   }
/* Check again the command line variables 
 * This will be uncommented when gaa is corrected.
 */

   if (gaa(argc, argv, &info) != -1) {
      err_quit(_
	       ("Error in the arguments. Use the --help parameter for more info.\n"));
   }
/* Examine how we were called */
   if (strstr(program_name, "decrypt") != NULL) {
      Bzero(program_name, strlen(program_name));
      strcpy(program_name, "mdecrypt");
      din = TRUE;
      ein = FALSE;
   } else {
      if (strlen(program_name) > 6) {
	 Bzero(program_name, strlen(program_name));
	 strcpy(program_name, "mcrypt");
      }
      din = FALSE;
      ein = TRUE;		/* It will change by the arguments */
   }



   /* File pointers are as much as the file arguments 
    */

   if ((info.size) * sizeof(char *) > 0) {
      file = _mcrypt_malloc((info.size) * sizeof(char *));
   } else {
      file = NULL;
   }

   if (info.ed_specified != 0) {
      din = info.din;
      ein = info.ein;
   }
   if (info.real_random_flag == TRUE) {
      real_random_flag = TRUE;
   }
   
   force = info.force;
   bare_flag = info.bare_flag;
   unlink_flag = info.unlink_flag;
   quiet = info.quiet;
   noecho = info.noecho;
   noiv = info.noiv;
   double_check = info.double_check;
   flush = info.flush;
   nodelete = info.nodelete;
   timer = info.timer;

   algorithms_directory = info.algorithms_directory;
   modes_directory = info.modes_directory;
#ifdef ZIP
   gzipflag = info.gzipflag;
   bzipflag = info.bzipflag;
#endif

   if (info.kmode != NULL) {
      keymode = info.kmode;
   } else {
      keymode = DEFAULT_KEYMODE;	/* Default */
   }

   if (info.hash != NULL) {
      hash = _mcrypt_strdup( info.hash);
      hash_algorithm = check_hash_algo(hash);

   } else {
      hash = _mcrypt_strdup( HASH_ALGORITHM);
      hash_algorithm = check_hash_algo(hash);
   }

   if (hash_algorithm < 0) {
      fprintf(stderr, _("The '%s' hash algorithm was not found.\n"), hash);
      fprintf(stderr,
	      _
	      ("Use the --hashlist parameter to view all supported hashes.\n"));
      exit(1);
   }

   if (info.keyfile != NULL) {
      kin = 2;
      keyfile = read_key_file(info.keyfile, &keys);
      if (keyfile == NULL)
	 kin = 0;
      if (keys == 0)
	 kin = 0;
   } else {
      if (info.keys != NULL) {
	 kin = 2;
	 keyfile = _mcrypt_malloc(info.keylen * sizeof(char *));
	 keys = info.keylen;

	 for (i = 0; i < info.keylen; i++) {
	    keyfile[i] = _secure_mcrypt_malloc(strlen(info.keys[i]) + 10);
	    strcpy(keyfile[i], info.keys[i]);
	 }
      }
   }

   if (info.mode != NULL) {
      mode = info.mode;
   } else {
      mode = DEFAULT_MODE;
   }

   openpgp = info.openpgp;
   if (openpgp!=0) {
   	mode = DEFAULT_PGP_MODE;
   	if (strncasecmp( "s2k", keymode, 3)!=0) {
   		keymode = DEFAULT_PGP_KEYMODE;
   	}
   }
   openpgp_z = info.openpgp_z;
   if ( openpgp_z < 0 || openpgp_z > 9 ) {
   	err_quit(_("Illegal compression level\n"));
   }

   if (info.algorithm != NULL) {
      used_algo = TRUE;
      algorithm = _mcrypt_strdup( info.algorithm);

      i = mcrypt_module_is_block_algorithm(algorithm,
					   algorithms_directory);

      if (i < 0) {
	 fprintf(stderr, _("Error in algorithm '%s'.\n"), algorithm);
	 exit(1);
      }

      if (i == 0) {
	 y = mcrypt_module_is_block_algorithm_mode(mode, modes_directory);
	 if (y >= 0) {
	    mode = "stream";
	 } else {
	    fprintf(stderr, _("Error in mode '%s'.\n"), mode);
	    exit(1);
	 }
      }

   } else {
      algorithm = DEFAULT_ALGORITHM;
   }

   if ( mode != NULL) {
      y = mcrypt_module_is_block_algorithm_mode(mode, modes_directory);
      if (y < 0) {
	 fprintf(stderr, _("Error in mode '%s'.\n"), mode);
	 exit(1);
      }
   }

   keysize = info.keysize;
   if ( keysize != 0) {
	check_keysize (keysize);

   } else
   	keysize = mcrypt_module_get_algo_key_size(algorithm, algorithms_directory);

   if (info.keylen != 0 && check_env() == FALSE) {
      err_warn(_
	       ("Warning: It is insecure to specify keywords in the command line\n"));
   }
#ifdef HAVE_UMASK
   umask(066);
#endif

/* For RAND32... Called only here */
#ifndef HAVE_DEV_RANDOM
      err_warn(_
	       ("Warning: This system does not have a random device. "
	       "Will use the included entropy gatherer.\n"));
#endif
   mcrypt_init_random();

/* Normal startup 
 */


/* ok now how many files were specified? */
   i = 0;

   /* '-' is used to refer to stdin */
   if (info.size == 1 && strcmp(info.input[0], "-") == 0)
      info.size -= 1;

   file_count += (info.size);

   if (file_count == 0) {
      stream_flag = TRUE;
   }
   while (i < info.size) {
      file[i] = info.input[i];
      i++;
   }


   if (stream_flag == TRUE)
      file_count++;

/* Do as many times as many files we got */
   for (i = 0; i < file_count; i++) {

      if (i != 0) {
	 if (outfile != NULL)
	    _mcrypt_free(outfile);
      }

      /* If keyword file specified choose the i-th keyword */

      if (kin == 2 && i <= (keys - 1)) {
	 keyword = keyfile[i];
	 if (i != 0) {
	    _secure_mcrypt_free(keyfile[i - 1], strlen(keyfile[i - 1]));	/* Free previous keyword */
	 }
      }
#ifdef HAVE_STAT
      if (stream_flag == FALSE) {
	 if (is_normal_file(file[i]) == FALSE) {
	    sprintf(tmperr,
		    _
		    ("%s: %s is not a regular file. Skipping...\n"),
		    program_name, file[i]);
	    err_crit(tmperr);
	    outfile = NULL;
	    continue;		/* next */
	 }
      }
#endif


      /* Check how we were called */
      if (din == TRUE) {	/* decryption */
      
	 if (stream_flag != TRUE)
	    dinfile = file[i];
	 if ((isatty(fileno((FILE *) (stdin))) == 1)
	     && (stream_flag == TRUE) && (force == 0)) {	/* not a tty */
	    sprintf(tmperr,
		    _
		    ("%s: Encrypted data will not be read from a terminal.\n"),
		    program_name);
	    err_crit(tmperr);
	    err_quit(_
		     ("Redirect the input instead.\nUse the --help parameter for more help.\n"));
	 }

         /* this will enable the pgp flag if it is a pgp file.
          * or disable it otherwise.
          */
         test_file_headers( dinfile);

      } else {			/* encryption */
	 if (stream_flag != TRUE)
	    einfile = file[i];
	 if ((isatty(fileno((FILE *) (stdout))) == 1)
	     && (stream_flag == TRUE) && (force == 0)) {	/* not a tty */
	    sprintf(tmperr,
		    _
		    ("%s: Encrypted data will not be written to a terminal.\n"),
		    program_name);
	    err_crit(tmperr);
	    err_quit(_
		     ("Redirect the output instead.\nUse the --help parameter for more help.\n"));
	 }
      }


      /* If no streams make the extensions */
      if (stream_flag != TRUE) {

	 if (din == TRUE) {
	 	outfile = remove_suffix( dinfile);
	 	if (outfile==NULL)
	 		continue;

	 } else {		/* encryption- append .nc */
	    outfile = _mcrypt_calloc(strlen(einfile) + 5 + 4, 1);
	    strcpy(outfile, einfile);
	    /* if file has already the .nc ignore it */
	    if (strstr(outfile, ".nc") != NULL) {
	       sprintf(tmperr,
		       _
		       ("%s: file %s has the .nc suffix... skipping...\n"),
		       program_name, outfile);
	       err_crit(tmperr);
	       continue;
	    }
#ifdef ZIP
	    if (stream_flag == FALSE) {
	       if (gzipflag == TRUE)
		  strcat(outfile, ".gz");
	       if (bzipflag == TRUE)
		  strcat(outfile, ".bz2");
	    }
#endif
	    strcat(outfile, ".nc");
#ifdef HAVE_STAT
	    /* But if it exists exit */
	    if (check_file(outfile) != 0) {
	       cleanDelete = FALSE;
	       if (ask_overwrite(program_name, outfile)
		   == FALSE) {
		  continue;
	       } else {
		  cleanDelete = TRUE;
	       }

	    } else {
	       cleanDelete = TRUE;
	    }
#endif

	 }
      } else {			/* if streams */
	 outfile = NULL;
      }



/* Decrypt */
      if (din == TRUE) {
	 if (openpgp!=0) x = pgp_decrypt_wrap( dinfile, outfile, keyword);
	 else x = decrypt_general(algorithm, dinfile, outfile, keyword);

	 if (x == 0) {
	    if (stream_flag == FALSE) {
	       sprintf(tmperr, _("File %s was decrypted.\n"), dinfile);
	       err_warn(tmperr);
	    } else {
	       sprintf(tmperr, _("Stdin was decrypted.\n"));
	       err_warn(tmperr);
	    }
#ifdef HAVE_STAT
# ifdef HAVE_UTIME_H
	    if (stream_flag == FALSE)
	       copyDate(dinfile, outfile);
# endif
#endif

	    if (unlink_flag == TRUE && stream_flag == FALSE)
	       x = unlink(dinfile);
	    if (x != 0)
	       perror("unlink");

	 } else {
	    if (stream_flag == FALSE) {
	       sprintf(tmperr,
		       _
		       ("File %s was NOT decrypted successfully.\n"),
		       dinfile);
	       err_crit(tmperr);
	       if (x != -1) {
		  if (nodelete == FALSE)
		     remove(outfile);
	       } else {
		  x = 1;
	       }
	    } else {
	       err_crit(_("Stdin was NOT decrypted successfully.\n"));
	    }
	 }

	 return_val += x;
      }

/* Encrypt */
      if (ein == TRUE) {
	 if (openpgp!=0) x = pgp_encrypt_wrap( einfile, outfile, keyword);
	 else x = encrypt_general(algorithm, einfile, outfile, keyword);

	 if (x == 0) {
	    if (stream_flag == FALSE) {
	       sprintf(tmperr, _("File %s was encrypted.\n"), einfile);
	       err_warn(tmperr);
	    } else {
	       sprintf(tmperr, _("Stdin was encrypted.\n"));
	       err_warn(tmperr);
	    }
#ifdef HAVE_STAT
#ifdef HAVE_UTIME_H
	    if (stream_flag == FALSE)
	       copyDate(einfile, outfile);
#endif
#endif
	    if (unlink_flag == TRUE && stream_flag == FALSE)
	       x = unlink(einfile);
	    if (x != 0)
	       perror("unlink");

	 } else {
	    if (stream_flag == FALSE) {
	       sprintf(tmperr,
		       _
		       ("File %s was NOT encrypted successfully.\n"),
		       einfile);
	       err_crit(tmperr);
	       if (x != -1) {
		  if (nodelete == FALSE)
		     remove(outfile);
	       } else {
		  x = 1;
	       }
	    } else {
	       err_crit(_("Stdin was NOT encrypted successfully.\n"));
	    }

	 }


	 return_val += x;
      }
   }				/* the main loop */

/* Clear the last keyword used */
   if (keyword != NULL) {
      Bzero(keyword, strlen(keyword));
      _mcrypt_free(keyword);
   }

/* Clear ALL arguments in the command line */
   for (i = 0; i < argc; i++) {
      Bzero(argv[i], strlen(argv[i]));
   }

   mcrypt_deinit_random();

   if (return_val != 0) {
      return 1;
   } else {
      return 0;
   }

}


int make_mult(int num, int d)
{
   int i = 0;

   while (i < num) {
      i += d;
   }
   return i;
}

static char* remove_suffix( char* filename) {
int y;
char* outfile;
	    
    y = strlen(filename);
    /* If the file has the .nc suffix, then remove it */

    if ((filename[y - 1] == 'c'
	 && filename[y - 2] == 'n' && filename[y - 3] == '.')) {

       outfile = _mcrypt_calloc(y - 2, 1);
       strncpy(outfile, filename, y - 3);

    } else /* gpg files */ if ((filename[y - 1] == 'g'
	 && filename[y - 2] == 'p' && filename[y - 3] == 'g'
	 && filename[y - 4] == '.')) {

	       	outfile = _mcrypt_calloc(y - 3, 1);
		strncpy(outfile, filename, y - 4);
    } else { /* append .dc */
       err_warn(_("Unknown suffix. Will append '.dc'.\n"));

       cleanDelete = TRUE;
       outfile = _mcrypt_calloc(y + 5, 1);
       strncpy(outfile, filename, y);
       strcat(outfile, ".dc");

    }

#ifdef HAVE_STAT
       /* But if it exists exit */
       if (check_file(outfile) != 0) {
	  cleanDelete = FALSE;
	  if (ask_overwrite( program_name, outfile) == FALSE) {
	     return NULL;
	  } else {
	     cleanDelete = TRUE;
	  }
       } else {
	  cleanDelete = TRUE;
       }
#endif

	return outfile;
}
